hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestContainerReplicationUnderPlacementPolicy.java

```diff
@@ -0,0 +1,178 @@
+    // needs at least two nodes to do remaining replica replication.
+
+    // case1: provide only 1 healthy candidate node, it should
+    // be failed to do replication.
+    verifyReplicateCommandCount(1, 0, null);
+
+    // case2: provide 2 healthy candidate node but one is chosen node
+    // that already be used, it should be also failed since chosen node
+    // will be excluded.
+    List<DatanodeDetails> chosenNodes = new ArrayList<>();
+    chosenNodes.add(randomDatanodeDetails());
+    verifyReplicateCommandCount(2, 0, chosenNodes);
+
+    // case3: provide other 2 healthy candidate node, it should
+    // be succeed to do replication.
+    verifyReplicateCommandCount(2, 2, null);
+  }
+
+  /**
+   * Check if replication behavior succeed by counting
+   * container replication command.
+   * @param numProvidedNode Number of provided datanode.
+   * @param expectedCount Expected replication command count.
+   * @param chosenNodes Provided node.
+   * @throws Exception
+   */
+  private void verifyReplicateCommandCount(int numProvidedNode,
+      int expectedCount, List<DatanodeDetails> chosenNodes) throws Exception {
+    List<DatanodeDetails> candidateNodes = new ArrayList<>();
+    if (chosenNodes != null) {
+      for (DatanodeDetails node : chosenNodes) {
+        candidateNodes.add(node);
+        numProvidedNode--;
+      }
+    }
+
+    for (int i = 0; i < numProvidedNode; i++) {
+      candidateNodes.add(randomDatanodeDetails());
+    }
+
+    conf = new OzoneConfiguration();
+    containerManager =
+        Mockito.mock(ContainerManager.class);
+    eventQueue = new EventQueue();
+    containerStateManager = new ContainerStateManager(conf);
+
+    Mockito.when(containerManager.getContainerIDs())
+        .thenAnswer(invocation -> containerStateManager.getAllContainerIDs());
+
+    Mockito.when(containerManager.getContainer(Mockito.any(ContainerID.class)))
+        .thenAnswer(invocation -> containerStateManager
+            .getContainer((ContainerID)invocation.getArguments()[0]));
+
+    Mockito.when(containerManager.getContainerReplicas(
+        Mockito.any(ContainerID.class)))
+        .thenAnswer(invocation -> containerStateManager
+            .getContainerReplicas((ContainerID)invocation.getArguments()[0]));
+
+    datanodeCommandHandler = new DatanodeCommandHandler();
+    eventQueue.addHandler(SCMEvents.DATANODE_COMMAND, datanodeCommandHandler);
+
+    SCMNodeMetric scmNodeMetric = Mockito.mock(SCMNodeMetric.class);
+    Mockito.when(scmNodeMetric.get()).thenReturn(new SCMNodeStat(1, 0, 1));
+
+    nodeManager =  Mockito.mock(NodeManager.class);
+    Mockito.when(nodeManager.getNodes(HddsProtos.NodeState.HEALTHY))
+        .thenReturn(candidateNodes);
+
+    Mockito.when(nodeManager.getNodeStat(Mockito.any(DatanodeDetails.class)))
+        .thenReturn(scmNodeMetric);
+
+    containerPlacementPolicy = new SCMContainerPlacementRandom(nodeManager,
+        conf, null, true, null);
```

Comment:
We can extract this to a method createAndConfigurePlacementPolicyWithMocks() method